### PR TITLE
Mc dropout reg

### PR DESCRIPTION
**Pull request by chelsealuisa**
- Added MC dropout model for regression
- Added 'final' (for poster/report) graphs for regression BNN, MLP and MC dropout

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -58,11 +60,11 @@ def reg_trainer():
     x_test = torch.linspace(-2., 2, config.num_test_points).reshape(-1, 1)
     for m_name, model in models.items():
         model.net.load_state_dict(torch.load(model.save_model_path, map_location=torch.device(DEVICE)))
-        y_test = model.evaluate(x_test, config.num_test_points)
```
Comment (tennisonliu) [inline main.py:61]: for my understanding 'config.num_test_points' became the attribute 'test_samples'?

File: regression/reg_task.py [hunk 2]
```diff
@@ -122,11 +123,69 @@ def train_step(self, train_data):
...
+        model_params = {
+            'input_shape': self.x_shape,
+            'classes': self.y_shape,
+            'batch_size': self.batch_size,
+            'hidden_units': self.hidden_units,
+            'mode': self.mode
+        }
+        self.net = MLP_Dropout(model_params).to(DEVICE)
+        self.optimiser = torch.optim.Adam(self.net.parameters(), lr=self.lr)
+        self.scheduler = torch.optim.lr_scheduler.StepLR(self.optimiser, step_size=5000, gamma=0.5)
```
Comment (tennisonliu) [inline regression/reg_task.py:165]: sorry to be pedantic, step_size=500 more apt given regression epochs

File: networks.py [hunk 3]
```diff
@@ -44,6 +44,12 @@ def sample(self):
     def log_prob(self, input):
         return (-math.log(math.sqrt(2 * math.pi)) - torch.log(self.sigma) - ((input - self.mu) ** 2) / (2 * self.sigma ** 2)).sum()
 
+def enable_dropout(model):
```
Comment (tennisonliu) [inline networks.py:47]: can this be a class method in MLP_Dropout?

**Comment by tennisonliu** [comment]
I am slightly unclear about what MC dropout actually does - the network is a regular (non-bayesian) MLP with dropout layers trained using vanilla SGD; the key difference being that sampling is enabled during evaluation to mimic the effects of dropout. during evaluation, each 'sample' of the network predictions is made with half the weights randomly zero-ed out. is this understanding correct?

File: main.py (hunk 1 above)
Comment (chelsealuisa) [reply main.py:61]: not quite: num_test_points is the number of x values that form the test set (see line 60), while test_samples is the number of predictions made for a given test point. I think we previously incorrectly had num_test_points here instead of test_samples (not that it should make much difference).

File: regression/reg_task.py (hunk 2 above)
Comment (chelsealuisa) [reply regression/reg_task.py:165]: thx, fixed!

File: networks.py (hunk 3 above)
Comment (chelsealuisa) [reply networks.py:47]: yep, good point.

**Comment by chelsealuisa** [comment]
yep, pretty much - all it does is to enable dropout during testing, so that with each prediction a random x% of the weights are dropped, which means that you can sample an average predictions in a similar way that we do for BNNs.

**Review by tennisonliu — APPROVED** [review]
